src/lib.rs

```diff
@@ -0,0 +1,136 @@
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::Mutex;
+
+const HOT_THRESHOLD: u16 = 10;
+
+/// What was the outcome of a `jit_control_point`?
+#[derive(Debug, PartialEq)]
+pub enum ControlOutcome {
+    /// The `jit_control_point` executed a trace: the interpreter does not need to execute a
+    /// trace for this location in the end user's code.
+    ExecutedTrace,
+    /// The `jit_control_point` did not execute a trace: the interpreter needs to execute the
+    /// relevant code for this location in the end user's code.
+    DidntExecuteTrace
+}
+
+/// The current meta-tracing state of a given location in the end-user's code.
+#[derive(Debug)]
+enum LocCount {
+    /// Below `HOT_THRESHOLD`, but counting up towards it.
+    Counting(u16),
+    /// A trace is being collected.
+    Tracing,
+    /// A compiled trace is available.
+    Trace
+}
+
+/// A meta-tracer.
+pub struct MetaTracer<Loc> {
+    counts: Mutex<HashMap<Loc, LocCount>>,
+    phantom: PhantomData<Loc>
+}
+
+impl<Loc: Eq + Hash> MetaTracer<Loc> {
+    pub fn new() -> Self {
+        Self {
+            counts: Mutex::new(HashMap::new()),
+            phantom: PhantomData
+        }
+    }
+
+    /// Attempt to execute a compiled trace for location `loc`: return a `ControlOutcome` to allow
+    /// the end user to determine whether a trace was executed or not.
+    pub fn control_point(&self, loc: Loc) -> ControlOutcome {
+        let mut cnts = self.counts.lock().unwrap();
+        let loc_cnt = cnts.entry(loc).or_insert(LocCount::Counting(0));
+        match *loc_cnt {
+            LocCount::Counting(x) => {
+                if x >= HOT_THRESHOLD {
+                    *loc_cnt = LocCount::Tracing;
+                } else {
+                    *loc_cnt = LocCount::Counting(x + 1);
+                }
+                ControlOutcome::DidntExecuteTrace
+            },
+            LocCount::Tracing => {
+                *loc_cnt = LocCount::Trace;
+                ControlOutcome::DidntExecuteTrace
+            },
+            LocCount::Trace => ControlOutcome::ExecutedTrace
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+    use std::thread;
+    use super::*;
+
+    #[test]
+    fn threshold_passed() {
+        let mt = MetaTracer::new();
+        for _ in 0..HOT_THRESHOLD {
+            assert_eq!(mt.control_point(0), ControlOutcome::DidntExecuteTrace);
+        }
+        assert_eq!(mt.control_point(0), ControlOutcome::DidntExecuteTrace);
+        assert_eq!(mt.control_point(0), ControlOutcome::DidntExecuteTrace);
+        assert_eq!(mt.control_point(0), ControlOutcome::ExecutedTrace);
+    }
+
+    #[test]
+    fn threaded_threshold_passed() {
+        let mt = Arc::new(MetaTracer::new());
+        let mut thrs = vec![];
+        for _ in 0..HOT_THRESHOLD {
+            let mt_cl = Arc::clone(&mt);
+            let t = thread::Builder::new()
```

Comment:
Can we re-introduce my named-thread utility? This means the threads can be given useful names which appear in backtraces and help debugging.

https://github.com/softdevteam/yorick_jit_runtime/blob/047f3204b59507f98738a55f03a529fe98865ce8/src/lib.rs#L287-L297

Reply:
This is a test, so I don't think it's relevant here. [It's a test that should not fail; if it does, naming threads is not the main concern!]

Reply:
I disagree on this one. If there is a concurrency bug causing a non-deterministic crash, we may not be able to easily reproduce it, and I'd like all the info I can get. Without my helper, all the test threads will have the same name.

Reply:
This is over-engineering in a test that isn't expected to fail. Why not preemptively add extra information that might be useful if a test fails? I could put a print statement between every line, for example. [Yes, that's a slightly facetious example, but there is a serious point here.]

Reply:
When I was working on #1, often test failures could only occurred on travis, and I found myself having to push debugging code, then wait for travis to crash (possibly re-starting the tests several times to reproduce the crash due to non-determinism), then repeat.

Rather than go through this dance every time, I think it would save us time to have diagnostics in place (as long as performance isn't impacted -- thread names should be fine). After all, it may be *users* finding the crashes. Do we want to ask them to apply patches and re-compile? Or would we rather they just post a back trace?

I'd suggest we name all our threads, in tests or otherwise.

Reply:
I don't think we're going to agree on this one -- sorry.

Reply:
So you don't think it would be useful to have threads named `compiling_loc_45` for example?

Reply:
I think we should do that for user visible threads. But there's no suggestion of a race in this test, so the extra support is bloat. And, the older I get, the more I think that bloat in code is our eternal enemy (literally: bloat never dies).